Add Maptool to handle chamfer fillet

benoitdm-oslandia opened the pull request:
This PR follows https://github.com/qgis/QGIS/pull/62572 and depends on https://github.com/qgis/QGIS/pull/63062

This PR adds a Maptool to chamfer or fillet polygons.

## Overview


https://github.com/user-attachments/assets/498bbeab-58d2-4f1a-811f-6e79f753aec9


Sponsored by: 
  * Stadt Frankfurt am Main 
  * Oslandia

benoitdm-oslandia committed:
feat: add map tool to handle chamfer fillet api

nyalldawson commented:
Can you add a screencast showing this in action?

benoitdm-oslandia commented:
> Can you add a screencast showing this in action?

Yes but for now it is only a draft it still need some improvements. Regards

nyalldawson commented:
Thanks @ptitjano !

The tool operation looks great, but I'm -1 to just throwing this into the digitizing toolbar. I'd much prefer it's exposed as a CAD dock tool.

uclaros commented:
> I'd much prefer it's exposed as a CAD dock tool.

I don't think it fits the context of the CAD dock well. It fits better where the offset and trim/extend tools are. Another option would be to integrate it with the vertex tool, but the vertex tool is already a complex mammoth...

I'd argue against the `ctrl` modifier modes, which seem to be modeled after the _offset_ tool. They cover some very specific use case workflow and are not generic enough to include in the tool itself.

A few more first thoughts:
Single chamfer distance mode (symmetric) should be enabled by default.
Entering of distance/radius should be keyboard-first for exact value entering. Mouse tracking should only be enabled after some user interaction (widget button maybe? shift-mouse move?)
The last used mode (chamfer/fillet) should be remembered and re-used on next tool run. (this will make mass fillets easier without needing a separate toolbar icon for fillet)
The last used radius/chamfer distance should be remembered for easy repetition.

benoitdm-oslandia commented:
> I don't think it fits the context of the CAD dock well. It fits better where the offset and trim/extend tools are. Another option would be to integrate it with the vertex tool, but the vertex tool is already a complex mammoth...
> 

Also I prefer when it is located near offset and trim/extend map tools.

> I'd argue against the `ctrl` modifier modes, which seem to be modeled after the _offset_ tool. They cover some very specific use case workflow and are not generic enough to include in the tool itself.

As you said, it is a copy paste of offset tool and the Ctrl mode is not necessary at all.

> 
> A few more first thoughts: Single chamfer distance mode (symmetric) should be enabled by default. Entering of distance/radius should be keyboard-first for exact value entering. 

Indeed keyboard input is the current default: the user clicks on a vertex, then enter the distance in the widget and validate with the enter key. Or the user can move the mouse and click for a second point.

The symmetric chamber can be obtain by using the Shift modifier (the doc is missing for now) during mouse operations.

> Mouse tracking should only be enabled after some user interaction (widget button maybe? shift-mouse move?) 

I am not fond of this idea.

> The last used mode (chamfer/fillet) should be remembered and re-used on next tool run. (this will make mass fillets easier without needing a separate toolbar icon for fillet) The last used radius/chamfer distance should be remembered for easy repetition.

For now only the last used mode and the number of segment for a fillet are saved. I also can add the chamfer distances and fillet radius.

baswein commented:
> > > Mouse tracking should only be enabled after some user interaction (widget button maybe? shift-mouse move?)
> > 
> > 
> > I am not fond of this idea.
> 
> Remembering last used radii/distances is not possible without it, as any mouse movement will overwrite the values. I'm pretty sure if you had to perform hundreds of `1.75r` fillets you would soon get fonder! AutoCAD is quite good at productivity and this is how they do it too

Many other tools in QGIS are: select and then click to set origin and click to set destination. Personally I would like to see this because it would allow you to reference other geometry and in the future I would like to be able to box or shift select multiple vertexes and them set all of them at once and I think this would work well for that.

benoitdm-oslandia committed:
fix(QgsMapToolChamferFillet): add buttons to lock values and improve repeatable tasks

benoitdm-oslandia commented:
> > > > Mouse tracking should only be enabled after some user interaction (widget button maybe? shift-mouse move?)
> > > 
> > > 
> > > I am not fond of this idea.
> > 
> > 
> > Remembering last used radii/distances is not possible without it, as any mouse movement will overwrite the values. I'm pretty sure if you had to perform hundreds of `1.75r` fillets you would soon get fonder! AutoCAD is quite good at productivity and this is how they do it too
> 
> Many other tools in QGIS are: select and then click to set origin and click to set destination. Personally I would like to see this because it would allow you to reference other geometry and in the future I would like to be able to box or shift select multiple vertexes and them set all of them at once and I think this would work well for that.

Fixed in last commit by adding button to lock values in widget.

nyalldawson commented:
Ok, given the discussion I'm removing my earlier -1 to the map tool approach.

uclaros commented:
Can you prepare this for review before marking it as ready for review? (remove unrelated changes, remove commented code, remove qDebug/qWarning, adjust level of QgsDebugMsgLevel etc)

benoitdm-oslandia commented:
rebase from master

benoitdm-oslandia commented:
rebase from master

lbartoletti committed:
fix(QgsGeometryUtils): ensure fillet tangent points are included in segmented geometry

The filletVertex function was skipping the exact tangent points when segmentizing
the fillet arc, causing misalignment with original segments especially with low
segment counts. Now explicitly adds tangent points to ensure proper alignment.

lbartoletti committed:
feat(QgsGeometryUtils): add maxFilletRadius utility function

Calculates the maximum allowed fillet radius for given segment configuration
to help constrain GUI interfaces and prevent exceeding geometric limits.

Includes comprehensive Python test coverage for various angle configurations.

lbartoletti committed:
test: update fillet tests to match corrected tangent point behavior

After fixing filletVertex to include exact tangent points in the geometry,
tests needed to be updated to reflect the correct output which now includes
these points.

lbartoletti committed:
feat(QgsMapToolChamferFillet): constrain fillet radius using maxFilletRadius

The fillet tool now automatically calculates and sets the maximum allowed
radius based on the selected vertex geometry, preventing users from entering
values that would exceed geometric constraints.

lbartoletti committed:
refactor(QgsGeometryUtilsBase): add maxFilletRadius to QgsGeometryUtilsBase and use it in QgsGeometryUtils

benoitdm-oslandia committed:
fixup! feat: add map tool to handle chamfer fillet api

nyalldawson commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,602 @@
+
+  // reduce the number of call to max 1 per 100ms
+  long currTime = QDateTime::currentMSecsSinceEpoch();
+  if ( currTime - mLastMouseMove < 100 )
```
Use a QElapsedTimer here instead

nyalldawson commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,602 @@
+  if ( !match.hasEdge() && !match.hasVertex() )
+    return;
+
+  mManipulatedGeometry = QgsGeometry( geom.constGet()->clone() );
```
```suggestion
  mManipulatedGeometry = geom;
```

nyalldawson replied:
Can you do some more familiarization with the QgsGeometry class please? It's implicitly shared, and already handles detachment for modifications internally.

nyalldawson commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,602 @@
+  {
+    op = "chamfer";
+    QgsSettingsRegistryCore::settingsDigitizingChamferFilletOperation->setValue( op );
+    qWarning() << "No op defined!! Default: chamfer";
```
cleanup

nyalldawson commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,602 @@
+
+  updateLabels( op );
+
+  // connect signals
+  connect( mOperationComboBox, static_cast<void ( QComboBox::* )( int )>( &QComboBox::currentIndexChanged ), this, //
```
remove the empty comments, and redundant "connect signals" comment

benoitdm-oslandia replied:
ok for "connect signals", but what are the other comments you mentioned?

nyalldawson replied:
here:

<img width="994" height="796" alt="image" src="https://github.com/user-attachments/assets/c32dc0ca-78ef-4eb8-a3fc-8921488e5743" />

benoitdm-oslandia replied:
code formatter generates a too long line without it. The comment increases the readability (IMPOV):

```cpp
  connect( mValue2SpinBox, static_cast<void ( QDoubleSpinBox::* )( double )>( &QDoubleSpinBox::valueChanged ), this, [this]( const double value ) {
    if ( operation() == QgsGeometry::Fillet )
      QgsSettingsRegistryCore::settingsDigitizingChamferFilletSegment->setValue( static_cast<int>( value ) );
    else
      QgsSettingsRegistryCore::settingsDigitizingChamferFilletValue2->setValue( value );
    emit distanceConfigChanged();
  } );
```

VS

```cpp
  connect( mValue2SpinBox, static_cast<void ( QDoubleSpinBox::* )( double )>( &QDoubleSpinBox::valueChanged ), this, //
           [this]( const double value ) {
             if ( operation() == QgsGeometry::Fillet )
               QgsSettingsRegistryCore::settingsDigitizingChamferFilletSegment->setValue( static_cast<int>( value ) );
             else
               QgsSettingsRegistryCore::settingsDigitizingChamferFilletValue2->setValue( value );
             emit distanceConfigChanged();
           } );
```

benoitdm-oslandia replied:
fixed

nyalldawson commented on src/app/qgsmaptoolchamferfillet.h:
```diff
@@ -0,0 +1,126 @@
+    std::unique_ptr<QgsSnapIndicator> mSnapIndicator;
+
+    //! The layer being manipulated
+    QgsVectorLayer *mSourceLayer = nullptr;
```
For safety this should be QPointer

benoitdm-oslandia replied:
ok but as QgsPointLocator::Match::layer() does not return a QPointer what we will gain?

benoitdm-oslandia committed:
fixup! feat: add map tool to handle chamfer fillet api

benoitdm-oslandia committed:
fixup! feat: add map tool to handle chamfer fillet api

benoitdm-oslandia commented on src/core/geometry/qgsgeometryutils_base.cpp:
```diff
@@ -739,6 +739,107 @@ bool QgsGeometryUtilsBase::bisector( double aX, double aY, double bX, double bY,
+
+  double dir1X, dir1Y, dir2X, dir2Y;
+
+  if ( dist1ToStart > epsilon )
+  {
+    dir1X = segment1StartX; dir1Y = segment1StartY;
+  }
+  else
+  {
+    dir1X = segment1EndX; dir1Y = segment1EndY;
+  }
```
ok fixed

benoitdm-oslandia commented on src/core/geometry/qgsgeometryutils_base.cpp:
```diff
@@ -739,6 +739,107 @@ bool QgsGeometryUtilsBase::bisector( double aX, double aY, double bX, double bY,
+    return -1.0;
+  }
+
+  const double maxRadius = maxDistanceToTangent * std::tan( halfAngle );
+
+  return maxRadius;
```
ok

benoitdm-oslandia committed:
fixup! refactor(QgsGeometryUtilsBase): add maxFilletRadius to QgsGeometryUtilsBase and use it in QgsGeometryUtils

benoitdm-oslandia commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,602 @@
+    return;
+  }
+
+  QgsVectorLayer *destLayer = qobject_cast<QgsVectorLayer *>( canvas()->currentLayer() );
```
ok

benoitdm-oslandia commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,602 @@
+    //get distance from current position rectangular to feature
+    const QgsPointXY layerCoords = toLayerCoordinates( mSourceLayer, mapPoint );
+
+    QgsVector vect = layerCoords - mVertexPoint;
```
If you prefer

benoitdm-oslandia committed:
fix(QgsMapToolChamferFillet): replace uses of operation strings by enums

benoitdm-oslandia commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,602 @@
+    QgsVector perpVect = vect.perpVector();
+
+    int beforeVIdx, afterVIdx;
+    mManipulatedGeometry.adjacentVertices( mVertexIndex, beforeVIdx, afterVIdx );
```
done

benoitdm-oslandia commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,602 @@
+
+    match = mCanvas->snappingUtils()->snapToCurrentLayer( e->pos(), QgsPointLocator::Types( QgsPointLocator::Vertex ) );
+
+    if ( auto *lLayer = match.layer() )
```
ok but why?

nyalldawson replied:
https://github.com/qgis/QGIS-Enhancement-Proposals/blob/master/qep-314-coding-style.md#3-language-features

benoitdm-oslandia committed:
fix(QgsMapToolChamferFillet): filter invalid input geometries

benoitdm-oslandia committed:
fix(QgsMapToolChamferFillet): avoid to generate duplicated points

nyalldawson commented on src/app/qgsmaptooloffsetcurve.h:
```diff
@@ -76,7 +76,7 @@ class APP_EXPORT QgsMapToolOffsetCurve : public QgsMapToolEdit
 
   private:
-    //! Rubberband that shows the position of the offset curve
+    //! Rubberband that shows the position of the offset  curve
```
not resolved yet (maybe you forgot to commit)

nyalldawson commented on src/core/geometry/qgsgeometryutils_base.cpp:
```diff
@@ -739,6 +739,107 @@ bool QgsGeometryUtilsBase::bisector( double aX, double aY, double bX, double bY,
+
+  if ( dist1ToStart > epsilon )
+  {
+    dir1X = segment1StartX; dir1Y = segment1StartY;
```
Actually it looks like all these aren't resolved -- please don't mark as resolved until the changes are actually in the PR.

nyalldawson commented on src/core/settings/qgssettingsregistrycore.h:
```diff
@@ -131,6 +131,24 @@ class CORE_EXPORT QgsSettingsRegistryCore : public QgsSettingsRegistry
     static const QgsSettingsEntryDouble *settingsDigitizingOffsetMiterLimit;
 
+    //! Settings entry digitizing chamfer/fillet: operation name
+    static const QgsSettingsEntryString *settingsDigitizingChamferFilletOperation;
```
> I will need to move QgsGeometry::ChamferFilletOperationType to qgis.h

Please do that, it's in keeping with current best-practice QGIS development approach

benoitdm-oslandia replied:
moved in qgsmaptoolchamferfillet.h

benoitdm-oslandia commented:
forgot to push...

benoitdm-oslandia committed:
fixup! feat: add map tool to handle chamfer fillet api

benoitdm-oslandia committed:
fix(QgsMapToolChamferFillet): refactor settings by replacing string by enum and migrate into maptool

nyalldawson commented:
@lbartoletti next time wait for my final review. I spent hours reviewing this pr so far, to merge without final approval given is bad form.

benoitdm-oslandia commented:
> @lbartoletti next time wait for my final review. I spent hours reviewing this pr so far, to merge without final approval given is bad form.

@nyalldawson  Sorry! I think it was more a mistake than a malicious act! I still can add some corrections if there are something you want to change?

uclaros commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,615 @@
+      mValue1SpinBox->setValue( value1 );
+
+      mVal2Label->setText( tr( "Distance 2" ) );
+      mValue2SpinBox->setDecimals( 6 );
```
6 digits? Do we really want micrometers here?

uclaros commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,615 @@
+
+      mVal2Label->setText( tr( "Distance 2" ) );
+      mValue2SpinBox->setDecimals( 6 );
+      mValue2SpinBox->setClearValue( 0.001 );
```
Why 0.001 ? One would expect either 0 or the default initial setting value

uclaros commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,615 @@
+
+      mVal2Label->setText( tr( "Fillet segments" ) );
+      mValue2SpinBox->setDecimals( 0 );
+      mValue2SpinBox->setClearValue( 6.0 );
```
min and max values should match those of the setting that will hold this value. Preferably stored on a constexpr so that they stay matched.

uclaros commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,615 @@
+
+void QgsMapToolChamferFillet::updateGeometryAndRubberBand( double value1, double value2 )
+{
+  if ( !mRubberBand || mOriginalGeometryInSourceLayerCrs.isNull() )
+  {
+    return;
+  }
+
+  if ( !mSourceLayer )
+  {
+    return;
+  }
```
group those two

uclaros commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,615 @@
+bool QgsMapToolChamferFillet::prepareGeometry( const QgsPointLocator::Match &match, QgsFeature &snappedFeature )
+{
+  const QgsVectorLayer *vl = match.layer();
+  if ( !vl )
```
`if ( !match.layer() )`

uclaros commented on src/ui/qgisapp.ui:
```diff
@@ -2258,6 +2260,27 @@ Click to select a geometry to offset.
+Creates a chamfer or fillet of a geometry:
+Click to select a geometry to operate.
+2nd click to specify the chamfer distance or fillet radius.
+Ctrl+2nd click to create a copy of the selected geometry.
+Ctrl+click with snapping to select another layer's geometry to copy and operate.
```
Does the ctrl modifier still apply??

benoitdm-oslandia replied:
see https://github.com/qgis/QGIS/pull/63527/files#diff-7ed5bdfb153b1620af72f740d8879928a6de308849cc8e043785c64d3d727b23

uclaros commented on src/app/qgsmaptoolchamferfillet.cpp:
```diff
@@ -0,0 +1,615 @@
+      i++;
+    }
+
+    emit messageEmitted( tr( "Generated geometry is not valid: '%1'. " ).arg( mModifiedGeometry.lastError() ) + message, Qgis::MessageLevel::Critical );
```
This is going to generate quite a big error message. Better add it to the log rather than the messagebar?

uclaros commented:
I've tried the feature on master and I'm genuinely curious if you are happy with the current usability state of the new tool.
There are several issues that need immediate attention before I'd consider it up to the (unwritten) QGIS quality standards.

First some critical issues that require immediate attention:

- Many crashes when clicking on a vertex
<details><summary>stack trace</summary>

```
src/gui/qgsmessagebar.cpp:318 : (pushItem) [11684ms] 2025-10-15T12:18:53 Messages[2] Chamfer or fillet : Creating chamfer/fillet geometry failed: Intersection 2 on segment but too far. Requested vertex: 1 was resolved as: [part: -1, ring: 0, vertex: 1]
src/gui/qgsmessagebar.cpp:318 : (pushItem) [6ms] 2025-10-15T12:18:53 Messages[2] Chamfer or fillet : Creating chamfer/fillet geometry failed: Intersection 2 on segment but too far. Requested vertex: 1 was resolved as: [part: -1, ring: 0, vertex: 1]
QGIS died on signal 11Extra Info File: /tmp/qgis-crash-info-148944
[New LWP 149042]
[New LWP 149008]
[New LWP 149007]
[New LWP 149004]
[New LWP 149003]
[New LWP 149002]
[New LWP 148986]
[New LWP 148977]
[New LWP 148976]
[New LWP 148975]
[New LWP 148974]
[New LWP 148973]
[New LWP 148972]
[New LWP 148971]
[New LWP 148970]
[New LWP 148969]
[New LWP 148968]
[New LWP 148967]
[New LWP 148966]
[New LWP 148965]
[New LWP 148964]
[New LWP 148963]
[New LWP 148962]
[New LWP 148961]
[New LWP 148960]
[New LWP 148959]
[New LWP 148958]
[New LWP 148954]
[New LWP 148953]
[New LWP 148952]
[New LWP 148951]
[New LWP 148950]
[New LWP 148949]
[New LWP 148948]
[New LWP 148946]
[New LWP 148945]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
__syscall_cancel_arch () at ../sysdeps/unix/sysv/linux/x86_64/syscall_cancel.S:56

warning: 56     ../sysdeps/unix/sysv/linux/x86_64/syscall_cancel.S: No such file or directory
[Current thread is 1 (Thread 0x7f4856017880 (LWP 148944))]
#0  __syscall_cancel_arch () at ../sysdeps/unix/sysv/linux/x86_64/syscall_cancel.S:56
No locals.
#1  0x00007f485f099668 in __internal_syscall_cancel (a1=<optimized out>, a2=<optimized out>, a3=<optimized out>, a4=<optimized out>, a5=a5@entry=0, a6=a6@entry=0, nr=61) at ./nptl/cancellation.c:49
        result = <optimized out>
        pd = <optimized out>
        ch = <optimized out>
#2  0x00007f485f0996ad in __syscall_cancel (a1=<optimized out>, a2=<optimized out>, a3=<optimized out>, a4=<optimized out>, a5=a5@entry=0, a6=a6@entry=0, nr=61) at ./nptl/cancellation.c:75
        r = <optimized out>
#3  0x00007f485f104787 in __GI___wait4 (pid=<optimized out>, stat_loc=<optimized out>, options=<optimized out>, usage=<optimized out>) at ../sysdeps/unix/sysv/linux/wait4.c:30
No locals.
#4  0x000055bf1ce669ac in qgisCrash (signal=11) at /home/uclaros/dev/cpp/QGIS/src/app/main.cpp:349
        status = 0
        pidstr = "--pid=148944", '\000' <repeats 19 times>
        gdbpid = 149069
        exename = "/home/uclaros/dev/cpp/QGIS/build-master-debug/output/bin/qgis", '\000' <repeats 450 times>
        len = 61
#5  <signal handler called>
No locals.
#6  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=11, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
        tid = <optimized out>
        ret = 0
        pd = <optimized out>
        old_mask = {__val = {0}}
        ret = <optimized out>
#7  0x00007f485f09e9ff in __pthread_kill_internal (threadid=<optimized out>, signo=11) at ./nptl/pthread_kill.c:89
No locals.
#8  0x00007f485f049cc2 in __GI_raise (sig=11) at ../sysdeps/posix/raise.c:26
        ret = <optimized out>
#9  <signal handler called>
No locals.
#10 0x00007f486b1159c4 in QObject::thread() const () from /lib/x86_64-linux-gnu/libQt5Core.so.5
No symbol table info available.
#11 0x00007f48682e5efe in QgsVectorLayer::wkbType (this=0x0) at /home/uclaros/dev/cpp/QGIS/src/core/vector/qgsvectorlayer.cpp:833
        __PRETTY_FUNCTION__ = "virtual Qgis::WkbType QgsVectorLayer::wkbType() const"
        __FUNCTION__ = "wkbType"
#12 0x00007f486f8442b5 in QgsMapToolChamferFillet::canvasReleaseEvent (this=0x55bf297a61d0, e=0x55bf2cffcee0) at /home/uclaros/dev/cpp/QGIS/src/app/qgsmaptoolchamferfillet.cpp:241
        hasZ = false
        hasM = false
        fet = {_vptr.QgsFeature = 0x7f48686fa020 <vtable for QgsFeature+16>, static staticMetaObject = {d = {superdata = {direct = 0x0}, stringdata = 0x7f4865aa6960 <qt_meta_stringdata_QgsFeature>, data = 0x7f4865aa6b80 <qt_meta_data_QgsFeature>, static_metacall = 0x7f48677c0e42 <QgsFeature::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, d = {d = 0x55bf28a498d0}}
        lLayer = 0x55bf2e0e3680
        match = {mType = QgsPointLocator::Vertex, mDist = 5.6829529673928629, mPoint = {mX = 224018.0796307697, mY = 5449561.5802525496, mIsEmpty = false}, mLayer = 0x55bf2e0e3680, mFid = -9, mVertexIndex = 1, mEdgePoints = {{mX = 0, mY = 0, mIsEmpty = true}, {mX = 0, mY = 0, mIsEmpty = true}}, mCentroid = {mX = 0, mY = 0, mIsEmpty = true}, mMiddleOfSegment = {mX = 0, mY = 0, mIsEmpty = true}}
```

</details>

- In Fillet mode, `Chamfer or fillet : Creating chamfer/fillet geometry failed: Intersection 1 on segment but too far. Requested vertex: 2 was resolved as: [part: -1, ring: 0, vertex: 2]` is constantly thrown on the messagebar when the radius does not fit the geometry.
    - The frequency of these messages is disruptive
    - The messages lack actionable information for the user
    - These messages are also logged as CRITICAL in the log
    - But more importantly, it finishes the operation, making it impossible for the user to move the mouse to the floating widget to change the settings.
 
- In Chamfer mode, when distances are locked, they only apply to the preview rubber band. The actual applied chamfer still uses the mouse location.
- In fillet, the total number of same-length segments should be what is defined in _Fillet Segments_.

Then some things that just degrade the user experience:

- `Chamfer or fillet : Could not find a nearby feature in any vector layer.` is spammed when clicked anywhere but on an active layer's feature's vertex. This was copied from the offset tool but is misleading in this case. It should be skipped altogether.
- Symmetric chamfer (possibly the most common chamfer case) is hard. One needs to lock both distances and type the distance in two separate spinboxes every time.
- Selecting the fillet radius with the mouse is not intuitive. Instead of using the arbitrary distance of mouse pointer to vertex position, it should be computing the radius needed so that the fillet rubberband is always under the mouse cursor. That would also allow snapping to be used at some point.
- Similar for the symmetric chamfer, mouse should track the rubber band.
- Selecting the asymmetric chamfer distances with the mouse is quite challenging and it triggers my [QWOP](https://en.wikipedia.org/wiki/QWOP) PTSD! Does it make any sense? Arbitrary chamfering can be better achieved using the reshape tool!

The less important bits that would be nice to haves:
- Tool should not allow clicking on linestring start/end points
- Symmetric chamfer should stay symmetric even when applied to two segments of different length. Now when the short segment is covered, the chamfere becomes asymmetric.
- Chamfer distances spinboxes are updated with bogus values when mouse distance gets bigger than the neighboring segments' lengths.

benoitdm-oslandia commented on src/core/geometry/qgsgeometryutils.h:
```diff
@@ -1383,6 +1383,22 @@ class CORE_EXPORT QgsGeometryUtils
+     *
+     * \since QGIS 4.0
+     */
+    static double maxFilletRadius( const QgsPoint &segment1Start, const QgsPoint &segment1End,
```
yes it is!
